### PR TITLE
Add support for breaking in the debugger

### PR DESCRIPTION
**Pull request by nre-ableton**
ping @mli-ableton @gck-ableton @toh-ableton

### CODE REVIEW

**Comment by toh-ableton** [comment]
I'm taking a look.

File: include/cucumber-cpp/internal/RegistrationMacros.hpp [hunk 1]
```diff
@@ -21,11 +21,18 @@
 // **************                 CUKE OBJECTS                 ************** //
 // ************************************************************************** //
 
+#if defined(CUKE_ENABLE_BREAKPOINT_SUPPORT)
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/RegistrationMacros.hpp]: Why do we need a macro to enable/disable the feature? Why can't it be always on?

File: src/BreakpointSupport.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+namespace cucumber {
+namespace internal {
+
+static bool shouldBreakIntoDebugger = false;
+
+CUKE_API_ void BreakpointSupport::shouldBreakIntoDebuggerInNextStep() {
+  shouldBreakIntoDebugger = true;
+}
+
+CUKE_API_ bool BreakpointSupport::checkAndClearShouldBreakIntoDebugger() {
```
Comment (toh-ableton) [inline src/BreakpointSupport.cpp]: What about 
```c++
CUKE_API_ bool BreakpointSupport::checkAndClearShouldBreakIntoDebugger() {
  const bool result = shouldBreakIntoDebugger;
  shouldBreakIntoDebugger = false;
  return result;
}
```

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#include <signal.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+
+#define CUKE_BREAK_INTO_DEBUGGER raise(SIGTRAP)
+
+#endif // COMPILER CHECK
+
+
+class CUKE_API_ BreakpointSupport {
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: What's the purpose of this class?

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "../CukeDll.hpp"
+
+namespace cucumber {
+namespace internal {
+
+#if defined(_MSC_VER)
+// The ", true" makes the expression to be of type bool, which is required
+// for the CUKE_BREAK_INTO_DEBUGGER macro
+#define CUKE_BREAK_INTO_DEBUGGER (__debugbreak(), true)
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: I don't understand. What requires the expression to be of type bool?

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+#elif defined(__clang__) && defined(__APPLE__) && !TARGET_OS_IPHONE
+#define CUKE_BREAK_INTO_DEBUGGER                                     \
+  _Pragma("clang diagnostic push")                                   \
+  _Pragma("clang diagnostic ignored \"-Wgnu-statement-expression\"") \
+    ({ asm("int3"); true; })                                         \
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Same here. Why do we need this strange gnu extension instead of just
```c++
  asm("int3");
```
❓

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace cucumber {
+namespace internal {
+
+#if defined(_MSC_VER)
+// The ", true" makes the expression to be of type bool, which is required
+// for the CUKE_BREAK_INTO_DEBUGGER macro
+#define CUKE_BREAK_INTO_DEBUGGER (__debugbreak(), true)
+
+#elif defined(__clang__) && defined(__APPLE__) && !TARGET_OS_IPHONE
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Isn't it rather 
```c++
#elif defined(__GNUC__) && (__i386 || __x86_64)
```
❓

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#define CUKE_BREAK_INTO_DEBUGGER                                     \
+  _Pragma("clang diagnostic push")                                   \
+  _Pragma("clang diagnostic ignored \"-Wgnu-statement-expression\"") \
+    ({ asm("int3"); true; })                                         \
+  _Pragma("clang diagnostic pop")
+
+#else // Generic Unix
+#include <signal.h>
+#include <unistd.h>
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Why do you need `unistd.h` and `sys/syscall.h`? You're only using `raise` as far as I can tell.

Comment (toh-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: It's also probably not such a good idea to `#include` from within
```c++
namespace cucumber {
namespace internal {
```
This will cause problems for everybody including these headers later. You should `#include` this at the top-level namespace.

File: include/cucumber-cpp/internal/step/StepMacros.hpp [hunk 8]
```diff
@@ -25,13 +26,45 @@ ::cucumber::internal::registerStep< step_name >(         \
...
+  }                                                                           \
+  void step_class_name ::bodyIfCompiledWithBreakpointSupport()
+/**/
+
+#define CUKE_STEP_WITH_BREAKPOINT_WITH_NAME_(step_name, step_matcher)         \
+  CUKE_STEP_WITH_NAME_(step_name, step_matcher)                               \
+  BODY_WITH_BREAKPOINT_IMPLEMENTATION_(step_name)
+/**/
+
+#define CUKE_STEP_WITH_BREAKPOINT_(step_matcher)                              \
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/StepMacros.hpp]: If we really need conditional breakpoint support (I don't think so), wouldn't it be nicer to just give two definitions of `CUKE_STEP_` instead of having this separate `CUKE_STEP_WITH_BREAKPOINT_ `?

**Review by toh-ableton — COMMENTED** [review]
It would be great if we could get rid of some of the complexity that comes with these macros. In particular, I question the value of having `CUKE_ENABLE_BREAKPOINT_SUPPORT`.

Thanks for adding back the feature.

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 4 above)
Comment (gck-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: The `CUKE_BREAK_INTO_DEBUGGER` implementation comes from ABL's assert implementation, where it was called in an expression context.  `__debugbreak()` though has a type of `void`, which leads to compiler errors.

The same applies to the clang branch below.  `asm` is a statement and can't be called in an expression context.

It seems though that breaking into the debugger here doesn't require the macro to behave as an expression.

File: include/cucumber-cpp/internal/RegistrationMacros.hpp (hunk 1 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/RegistrationMacros.hpp]: I was anticipating that this would make merging it into the public cucumber-cpp easier.

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 4 above)
Comment (toh-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Yes, but we're not using it in an expression context here, right?

File: include/cucumber-cpp/internal/RegistrationMacros.hpp (hunk 1 above)
Comment (toh-ableton) [reply include/cucumber-cpp/internal/RegistrationMacros.hpp]: Why? 

From my point of view, there's lots of added complexity but no benefit. Am I missing something?

Comment (nre-ableton) [reply include/cucumber-cpp/internal/RegistrationMacros.hpp]: No, you're probably right, I'll remove the option.

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 4 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: No, we aren't. I'll remove the comment and change the macro to just be `__debugbreak()`. Should I also change the mac part to be just `{ asm("int3"); }`?

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 6 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: I'm not 100% sure. Won't the clang pragmas cause issues with other compilers? IIRC gcc's `-Wall` also includes `-Wunknown-pragma` (I might be mistaken here).

Comment (toh-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: I don't think the pragmas are necessary once you get rid of the non-standard statement expression.

File: src/BreakpointSupport.cpp (hunk 2 above)
Comment (nre-ableton) [reply src/BreakpointSupport.cpp]: Sure, works for me.

**Comment by mli-ableton** [comment]
@nre-ableton What's the status of this PR? Just asking, no rush. I just happened to need the feature today again.

**Comment by nre-ableton** [comment]
@mli-ableton Sorry, I've been busy with work on the new CI stuff and a few other PRs to cucumber-cpp. But this one has not been forgotten. 😄 Hopefully I'll have time to dust it off and work on it in the upcoming sprint.

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 4 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Fixed

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 6 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Ok, fixed.

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 5 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Not sure, but removed.

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 7 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Done

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 3 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: It was just to serve as a container for the two functions. However it's not strictly necessary. I've removed it in the interest of keeping this patch small.

File: include/cucumber-cpp/internal/step/StepMacros.hpp (hunk 8 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/StepMacros.hpp]: Hmm, interesting point. I'll remove the other `CUKE_STEP_` definition.

**Comment by nre-ableton** [comment]
So I've pushed an updated version of this branch, just so everybody can take a look, but it shouldn't be merged yet because we need to land the shared library support into cucumber-cpp upstream first. Also the `WIP: step matching hook` commit (1dd9fcb) is only necessary to compile this patch with Live, obviously this commit should not be in this branch otherwise.

File: CMakeLists.txt [hunk 9]
```diff
@@ -4,6 +4,7 @@ project(Cucumber-Cpp)
 
 set(CUKE_USE_STATIC_BOOST ${WIN32} CACHE BOOL "Statically link Boost (except boost::test)")
 set(CUKE_USE_STATIC_GTEST ON CACHE BOOL "Statically link Google Test")
+set(CUKE_ENABLE_SHARED_LIB OFF CACHE BOOL "Generate a shared library")
```
Comment (toh-ableton) [inline CMakeLists.txt:7]: The commit message has a few issues:
> On Windows, it is necessary to add some __declspec() calls to expose
some symbols with different linkage types. This commit adds the makes
it possible for users to easily build cucumber-cpp either as a static
or dynamic library by defining either CUKE_LINKED_AS_SHARED_LIBRARY or
CUKE_CREATE_SHARED_LIBRARY at build time.

* `__declspec(...)` is an attribute, not a function call.
* you don't use it "to expose some symbols with different linkage types". `__declspec(dllexport)` tells the compiler that the function, data or class should be exported when linking the resulting object file into a dll. `__declspec(dllimport)` conversely tells it that the respective definition is in a dll and needs to be resolved by the dynamic linker when loading the dll.
* `This commit adds the makes it possible` doesn't make sense gramatically.
* The sentence is also wrong. When building statically, neither of these macros should be defined. When building dynamically, you need to define `CUKE_CREATE_SHARED_LIBRARY` when building the dll and `CUKE_LINKED_AS_SHARED_LIBRARY` when using the dll.
* Looking further, it seems you're using a different mechanism provided by CMake that relies on a generated `CukeExport.hpp` header. Is this a request from upstream? It means that we'd have to maintain one more file in `build-system` since we're not using CMake. In case this doesn't come from upstream, I'd prefer the no header solution that we're using already.

File: CMakeLists.txt [hunk 10]
```diff
@@ -50,6 +51,8 @@ if(NOT CUKE_DISABLE_BOOST_TEST)
     set(Boost_USE_STATIC_LIBS OFF)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DBOOST_TEST_DYN_LINK")
     find_package(Boost ${BOOST_MIN_VERSION} COMPONENTS unit_test_framework)
+    # Save the unit test library in Boost_LIBRARIES for use in tests/CMakeLists.txt
```
Comment (toh-ableton) [inline CMakeLists.txt:54]: I don't understand this change. Can you elaborate, please?

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -33,8 +35,25 @@ if(CMAKE_EXTRA_GENERATOR OR MSVC_IDE)
...
-add_library(cucumber-cpp-nomain STATIC ${CUKE_SOURCES})
-add_library(cucumber-cpp STATIC ${CUKE_SOURCES} main.cpp)
+if(CUKE_ENABLE_SHARED_LIB)
+    set(CUKE_LINKAGE SHARED)
+else()
+    set(CUKE_LINKAGE STATIC)
+endif()
+
+add_library(cucumber-cpp-nomain ${CUKE_LINKAGE} ${CUKE_SOURCES})
+add_library(cucumber-cpp ${CUKE_LINKAGE} ${CUKE_SOURCES} main.cpp)
```
Comment (toh-ableton) [inline src/CMakeLists.txt:46]: I think that this is questionable. Why would you want to have your `main()` in a dll?

File: src/CMakeLists.txt [hunk 12]
```diff
@@ -33,8 +35,25 @@ if(CMAKE_EXTRA_GENERATOR OR MSVC_IDE)
...
+
+set_target_properties(cucumber-cpp cucumber-cpp-nomain PROPERTIES
```
Comment (toh-ableton) [inline src/CMakeLists.txt:48]: I'm assuming this is used to distinguish between building and using the dll, correct?

File: src/CMakeLists.txt [hunk 13]
```diff
@@ -33,8 +35,25 @@ if(CMAKE_EXTRA_GENERATOR OR MSVC_IDE)
...
+    DEFINE_SYMBOL cucumber_cpp_EXPORTS
+)
+
+generate_export_header(cucumber-cpp
```
Comment (toh-ableton) [inline src/CMakeLists.txt:52]: Why do you generate an export header only for `cucumber-cpp`, but not `cucumber-cpp-nomain`? See also my other comment questioning the usefulness of having `main` in a dll.

File: include/cucumber-cpp/internal/CukeEngine.hpp [hunk 14]
```diff
@@ -95,7 +97,8 @@ class CukeEngine {
      */
     virtual std::string snippetText(const std::string & keyword, const std::string & name, const std::string & multilineArgClass) const = 0;
 
-    virtual ~CukeEngine() {}
+    CUCUMBER_CPP_EXPORT CukeEngine();
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/CukeEngine.hpp:100]: Why did you add these? They aren't necessary in similar situations in the Live codebase.

File: src/CMakeLists.txt [hunk 15]
```diff
@@ -33,8 +35,25 @@ if(CMAKE_EXTRA_GENERATOR OR MSVC_IDE)
...
+    EXPORT_FILE_NAME "CukeExport.hpp"
+)
+
+target_include_directories(cucumber-cpp-nomain PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
```
Comment (toh-ableton) [inline src/CMakeLists.txt:56]: Shouldn't this be `${CMAKE_BINARY_DIR}/exports` instead?

File: include/cucumber-cpp/internal/hook/HookRegistrar.hpp [hunk 16]
```diff
@@ -64,6 +64,7 @@ class CUCUMBER_CPP_EXPORT HookRegistrar {
 public:
     typedef std::list< boost::shared_ptr<Hook> > hook_list_type;
     typedef std::list< boost::shared_ptr<AroundStepHook> > aroundhook_list_type;
+    typedef bool(*StepMatchingHook)(const boost::cmatch&);
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/hook/HookRegistrar.hpp:67]: Can you elaborate what's WIP about this commit?

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp [hunk 17]
```diff
@@ -0,0 +1,29 @@
+#ifndef CUKE_BREAKPOINTSUPPORT_HPP_
+#define CUKE_BREAKPOINTSUPPORT_HPP_
+
+#include "../CukeDll.hpp"
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: This file doesn't exist. Did you try compiling your code?

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <signal.h>
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: Are you sure that this header is available on all supported platforms?

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+#elif defined(__GNUC__) && (__i386 || __x86_64)
+#define CUKE_BREAK_INTO_DEBUGGER asm("int3")
+
+#else // Generic Unix
+#define CUKE_BREAK_INTO_DEBUGGER raise(SIGTRAP)
+
+#endif // COMPILER CHECK
+
+
+CUKE_API_ void shouldBreakIntoDebuggerInNextStep();
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: This won't work with your changes to the previous commit.

File: include/cucumber-cpp/internal/step/StepMacros.hpp [hunk 20]
```diff
@@ -25,13 +22,36 @@ ::cucumber::internal::registerStep< step_name >(         \
...
+/**/
+
 // ************************************************************************** //
 // **************               GIVEN/WHEN/THEN                ************** //
 // ************************************************************************** //
 
-#define GIVEN CUKE_STEP_
-#define WHEN CUKE_STEP_
-#define THEN CUKE_STEP_
+#define GIVEN CUKE_STEP_WITH_BREAKPOINT_
```
Comment (toh-ableton) [inline include/cucumber-cpp/internal/step/StepMacros.hpp]: Again, did you check that this even compiles? I don't see a definition for this macro, and I don't think you'd need to change it.

**Review by toh-ableton — COMMENTED** [review]
I think there are a few issues that you need to resolve before we can merge this.

The first one is the base-branch for this PR. I don't think that `master` is the right one. IMHO, `master` should always track upstream; and we should have an `ableton-master`-branch or similar where we maintain our patches.

I also don't quite understand your comment about the `WIP: step matching hook`-commit. What do you plan on doing about this? 

The code also contains a mixture of approaches for handling dynamic linkage. If the cmake module isn't a requirement from upstream, I'd favor a simple solution like we had before because it makes it easier to integrate with other build systems such as Live's.

Lastly, it seems to me that the code cannot compile in its current state. It would be great if you tried to make sure that your code compiles and works as expected. Do you have a corresponding branch in https://github.com/AbletonAppDev/live?

File: include/cucumber-cpp/internal/step/BreakpointSupport.hpp (hunk 18 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/step/BreakpointSupport.hpp]: It built on Windows, so I assume it'll be fine on all other platforms we care about...

**Comment by nre-ableton** [comment]
@toh-ableton as I tried (unsuccessfully) to explain in my previous comment, the commits 1dd9fcb and 48468ab shouldn't be part of this PR, but I cherry picked them in this branch temporarily just to get it building. I tested this code on my Windows workstation and got it building there, but perhaps I forgot to push something since you are correct about the macros and some other things not compiling correctly. I'll check that out once I'm back on my desktop and re-push.

Regarding the overall merging strategy for cucumber-cpp, we had a discussion in Gotham about what to do, since Live currently points to `nre_master_ableton-octo` which is not ideal. We decided that the plan of action will be to create a tag for that branch's `HEAD`, and then to point Live to `master` instead. We will maintain our own `master` (it is a fork, after all) and try to update it regularly from upstream and to contribute as many of our patches upstream as we are able. If you have any concerns about this, please voice them now, since we have decided to do this work only after landing our patches that are currently in review on the public cucumber-cpp project.

Regarding commit 1dd9fcb, that is marked as WIP since we did not figure out exactly what to do with that commit before attempting to upstream it in https://github.com/cucumber/cucumber-cpp/pull/149. What we should do with that PR is probably a separate discussion, but we probably should make the necessary changes discussed in that PR, drop the WIP commit, and also close the PR in cucumber-cpp.

As for 48468ab, that is part of a PR here: https://github.com/cucumber/cucumber-cpp/pull/147. I'm still trying to get everything building and passing on CI after the upstream maintainers requested changes. However, I'll look through some of your line comments and apply them to that branch.

File: CMakeLists.txt (hunk 9 above)
Comment (nre-ableton) [reply CMakeLists.txt:7]: Thanks for the comments on this admittedly mangled commit message. 😉 I'll fix and re-push in that branch. However, I just wanted to respond to the `CukeExport.hpp` part -- yes, this was a request from the upstream maintainers. Unfortunately you are correct, this does mean that we'll need to maintain such a file in `build-system`. I've got a branch there with this file (and the necessary GYP changes as well), the branch is named `nre_master_cucumber-cpp-includes`, and I will make a PR for it once we land the static/shared patch in cucumber-cpp's upstream.

File: CMakeLists.txt (hunk 10 above)
Comment (nre-ableton) [reply CMakeLists.txt:54]: Because the project's `CMakeLists.txt` calls `find_package(Boost ...)` multiple times, `Boost_LIBRARIES` can't be used to figure out where the Boost unit test framework is. Since cucumber-cpp always links this library dynamically, we need to copy it to the build dir when we are building a shared library. Otherwise, there are a bunch of runtime errors which occur on Windows because it won't be able to find that DLL when running the tests.

File: src/CMakeLists.txt (hunk 11 above)
Comment (nre-ableton) [reply src/CMakeLists.txt:46]: Yeah, I felt the same way, but this is how they do it. I didn't want to change the build more than is necessary in that respect.

File: src/CMakeLists.txt (hunk 12 above)
Comment (nre-ableton) [reply src/CMakeLists.txt:48]: I'm actually not sure why this is here anymore; IIRC it was added by request of the upstream maintainers.

File: src/CMakeLists.txt (hunk 13 above)
Comment (nre-ableton) [reply src/CMakeLists.txt:52]: Good point, I added another `generate_export_header` for `cucumber-cpp-nomain`.

File: src/CMakeLists.txt (hunk 15 above)
Comment (nre-ableton) [reply src/CMakeLists.txt:56]: No, as far as I can tell, `CukeExport.hpp` is generated in `build/src`

File: include/cucumber-cpp/internal/CukeEngine.hpp (hunk 14 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/CukeEngine.hpp:100]: It was at the request of the upstream maintainers

File: include/cucumber-cpp/internal/hook/HookRegistrar.hpp (hunk 16 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/hook/HookRegistrar.hpp:67]: See my big comment below

**Comment by nre-ableton** [comment]
@toh-ableton Ok, I spent some time dealing with the linkage issues and got it building (and working), at least on Windows. I'll give Mac a test tomorrow. Sorry for pushing an incomplete commit earlier today; I guess I thought I was much closer to a working commit than I actually was.

Also, if you are interested, the branch in Live for testing this commit is `nre_master_and-break`.

File: CMakeLists.txt (hunk 9 above)
Comment (toh-ableton) [reply CMakeLists.txt:7]: I've had a look at the branch in build-system. If you take a look at the `CukeExport.hpp`, you'll see that we only need one, not one for static and one for dynamic linking. In static builds, it's sufficient to set `CUCUMBER_CPP_STATIC_DEFINE` from the `gypi`.

**Comment by toh-ableton** [comment]
I think it's confusing to have a `master` branch that doesn't track upstream. On the other hand, if that's what we do for other forks, I'm fine with it.

**Comment by ala-ableton** [comment]
@toh-ableton: if our `master` branch tracks upstream, then either `live.git` never points to it (what's the point of having a `master` branch at all?), or we have to wait for our PRs to land upstream before being able to use them. I find that more confusing.

**Comment by toh-ableton** [comment]
@ala-ableton I don't understand why live needs to point to `master`, and I don't think it's a problem if it doesn't. In the submodule config, you can set a branch that should be used to update the submodule, and this doesn't necessarily have to be master.

**Comment by ala-ableton** [comment]
@toh-ableton: then let's delete our `master` branch. I don't see why we should keep a branch that we don't use.

**Comment by toh-ableton** [comment]
I think it's nice to see the difference of where our branch is in relation to upstream master. My intuition is that a fork should be as close as possible to the original repo. But maybe this just my thinking.

**Comment by ala-ableton** [comment]
If our `master` is identical to upstream's `master`, then there is no difference to see, but I think I am missing something. I agree that a fork should be as close as possible to the original repo, and this is usually done by merging upstream's `master` into the fork's `master` as regularly as possible.

**Comment by toh-ableton** [comment]
If I see a branch called `master`, I expect this to be upstream. I wouldn't realize immediately that there are commits that are not upstreamed if they're on a branch called `master`.

For the same reason, I don't develop on the `master`-branch on my machine. I could develop my features on my local `master`, and then `git pull` would merge `origin/master` into my `master` (or rather rebase). Maybe some people like to work that way, but I like to work on branches that I only merge into my local master when I'm ready to push this merge to origin. After you `git pull` in the `master` branch in my working copies, `master` always points to `origin/master`. 

I guess in my mental model, a branch (including `master`) lives in a global namespace, and you can always use `git push` and `git pull` to synchronize them. So if there is a branch (anywhere) that you can't synchronize with some other branch (such as upstream `master`) because one contains commits that cannot (yet) be merged into the other, I think they should have a different name.

**Comment by ala-ableton** [comment]
I personally don't have a local `master` branch in any of my working copies, except in repositories where I work directly on it (for instance `release-descriptions`), which is why I don't see the point of keeping a `master` branch that is identical to upstream.

You are right that it is easy to track another branch than `master` in a submodule, but when I am cloning `AbletonAppDev/cuucumber-cpp` to work on it, I am expecting to find *our* state, not the upstream state.

**Comment by nre-ableton** [comment]
@ala-ableton @toh-ableton This PR is really, really old. At this point I am not entirely sure what needs to happen with this branch. I would like to close this PR, but I'll keep the remote branch around in case we need to revive the work at some point. Any objections?

**Comment by ala-ableton** [comment]
@nre-ableton no objections!
